sled-agent/src/swap_device.rs

```diff
@@ -0,0 +1,504 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+/// Operations for creating a system swap device.
+use std::io::Read;
+use zeroize::Zeroize;
+
+#[derive(Debug, thiserror::Error)]
+pub enum SwapDeviceError {
+    #[error("Could not find boot disk")]
+    BootDiskNotFound,
+
+    #[error("Error running ZFS command: {0}")]
+    Zfs(illumos_utils::ExecutionError),
+
+    #[error("Error listing swap devices: {0}")]
+    ListDevices(String),
+
+    #[error("Error adding swap device: {msg} (path=\"{path}\", start={start}, length={length})")]
+    AddDevice { msg: String, path: String, start: u64, length: u64 },
+
+    #[error("{msg}: {error}")]
+    Misc { msg: String, error: String },
+}
+
+/// Ensure the system has a swap device, creating the underlying block
+/// device if necessary.
+///
+/// The swap device is an encrypted zvol that lives on the M.2 disk that the
+/// system booted from.  Because it booted from the disk, we know for certain
+/// the system can access it. We encrypt the zvol because arbitrary system
+/// memory could exist in swap, including sensitive data. The zvol is encrypted
```

Comment:
Hm, I forget what exactly we decided here. I see that @faithanalog is assigned to #2450. Do you have any context there?

Otherwise, I need to see if I can find the recording of the meeting we had for the design of the swap device and see if we discussed it there.

Reply:
Yeah, definitely no action needed from this PR, but mostly just a musing, since dump and swap are comparable for "what memory is written to disk".